File: examples/node/iam-create-policy.js
// Crea una política IAM de solo lectura sobre un bucket S3 específico
// (principio de mínimo privilegio, no AdministratorAccess).
// Uso: node iam-create-policy.js <nombre-politica> <nombre-bucket>
const AWS = require('aws-sdk');

const iam = new AWS.IAM({
  endpoint: 'http://localhost:4566',
  region: 'us-east-1',
  credentials: { accessKeyId: 'test', secretAccessKey: 'test' },
});

const [, , policyName, bucketName] = process.argv;

async function main() {
  if (!policyName || !bucketName) {
    throw new Error('Uso: node iam-create-policy.js <nombre-politica> <nombre-bucket>');
  }

  const document = {
    Version: '2012-10-17',
    Statement: [
      {
        Effect: 'Allow',
        Action: ['s3:GetObject', 's3:ListBucket'],
        Resource: [`arn:aws:s3:::${bucketName}`, `arn:aws:s3:::${bucketName}/*`],
      },
    ],
  };

  const result = await iam.createPolicy({
    PolicyName: policyName,
    PolicyDocument: JSON.stringify(document),
  }).promise();

  console.log('Política creada:', result.Policy.PolicyName);
  console.log('ARN:', result.Policy.Arn);
  console.log('Siguiente paso: node iam-attach-policy.js <usuario>', result.Policy.Arn);
}

main().catch((error) => {
  console.error('Error creando la política:', error.message);
  process.exit(1);
});

File: examples/python/iam_create_policy.py
"""Crea una política IAM de solo lectura sobre un bucket S3 específico
(principio de mínimo privilegio, no AdministratorAccess).

Uso: python3 iam_create_policy.py <nombre-politica> <nombre-bucket>
"""
import json
import sys

import boto3

iam = boto3.client(
    "iam",
    endpoint_url="http://localhost:4566",
    aws_access_key_id="test",
    aws_secret_access_key="test",
    region_name="us-east-1",
)


def main():
    if len(sys.argv) < 3:
        raise SystemExit("Uso: python3 iam_create_policy.py <nombre-politica> <nombre-bucket>")
    policy_name, bucket_name = sys.argv[1], sys.argv[2]

    document = {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Action": ["s3:GetObject", "s3:ListBucket"],
                "Resource": [f"arn:aws:s3:::{bucket_name}", f"arn:aws:s3:::{bucket_name}/*"],
            }
        ],
    }

    result = iam.create_policy(PolicyName=policy_name, PolicyDocument=json.dumps(document))
    print("Política creada:", result["Policy"]["PolicyName"])
    print("ARN:", result["Policy"]["Arn"])
    print("Siguiente paso: python3 iam_attach_policy.py <usuario>", result["Policy"]["Arn"])


if __name__ == "__main__":
    main()
